### PR TITLE
Update the version of resource after updates are downloaded from the server

### LINKED ISSUE #2271: Intermittent patches are being updated.
**Describe the bug**

On the Device:

1. Disable the device's internet connection.
   
2. Update the patient's phone number on the device.
   
3. Save the changes on the device.

On the Test Server:

4. Access the same patient record for which the phone number was updated on the device.

5. Update the patient's phone number with a different value on the server.

6. Save the patient record on the server.

Back on the Device:

7. Re-enable the internet connection on the device.

8. Trigger a manual synchronization between the device and the server.

Outcome:

9. Check whether the changes made on the device and the server are matching or not. 

Expectation:

10. The expectation is that the updated phone number for the patient on the device and server should be in sync, and both should show the same, updated version of the patient record.

**Component**
Core library, 

**To Reproduce**

**Expected behavior**

**Screenshots**

**Smartphone (please complete the following information):**
 - Device: Nokia 3.4, Emulator
 - Android version: android 12
 - Build number: [e.g. Settings -> About phone -> Build number]

**Additional context**

**Would you like to work on the issue?**
Please state if this issue should be assigned to you or who you think could help to solve this issue.

### PR DESCRIPTION
**Pull request by aditya-07**
Fixes #2271

**Description**
When updates are downloaded from the server, the database update code wasn't putting the new versionID of the updated resource into the table. 

**Alternative(s) considered**
Have you considered any alternatives? And if so, why have you chosen the approach in this PR?

**Type**
Choose one: Bug fix 
**Screenshots (if applicable)**

**Checklist**
- [x] I have read and acknowledged the [Code of conduct](https://github.com/google/android-fhir/blob/master/CODE_OF_CONDUCT.md).
- [x] I have read the [Contributing](https://github.com/google/android-fhir/wiki/Contributing) page.
- [x] I have signed the Google [Individual CLA](https://cla.developers.google.com/about/google-individual), or I am covered by my company's [Corporate CLA](https://cla.developers.google.com/about/google-corporate ).
- [x] I have discussed my proposed solution with code owners in the linked issue(s) and we have agreed upon the general approach.
- [x] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the style guide of this project.
- [x] I have run `./gradlew check` and `./gradlew connectedCheck` to test my changes locally.
- [x] I have built and run the demo app(s) to verify my change fixes the issue and/or does not break the demo app(s).

### CODE REVIEW

File: engine/src/main/java/com/google/android/fhir/MoreResources.kt [hunk 1]
```diff
@@ -57,7 +57,7 @@ fun <R : Resource> getResourceClass(resourceType: String): Class<R> {
   return Class.forName(R4_RESOURCE_PACKAGE_PREFIX + className) as Class<R>
 }
 
-internal val Resource.versionId
+internal val Resource.versionId: String?
```
Comment (jingtang10) [inline engine/src/main/java/com/google/android/fhir/MoreResources.kt:60]: why is this change needed? something to do with java/kotlin interoperability?

File: engine/src/main/java/com/google/android/fhir/db/impl/dao/ResourceDao.kt [hunk 2]
```diff
@@ -67,6 +67,7 @@ internal abstract class ResourceDao {
           serializedResource = iParser.encodeResourceToString(resource),
           lastUpdatedLocal = timeOfLocalChange,
           lastUpdatedRemote = lastUpdatedRemote?.toInstant() ?: it.lastUpdatedRemote,
+          versionId = resource.versionId ?: it.versionId,
```
Comment (jingtang10) [inline engine/src/main/java/com/google/android/fhir/db/impl/dao/ResourceDao.kt:70]: please add a test

**Review by jingtang10 — CHANGES_REQUESTED** [review]

File: engine/src/main/java/com/google/android/fhir/db/impl/dao/ResourceDao.kt (hunk 2 above)
Comment (aditya-07) [reply engine/src/main/java/com/google/android/fhir/db/impl/dao/ResourceDao.kt]: Done. 
Also, after our discussion I realised that having separate apis for local and remote updates would simplify code readability and correctness incase app developers add updated meta data to the locally changed resource that ideally should be ignored by engine as those values are assigned by server only.

File: engine/src/test/java/com/google/android/fhir/impl/FhirEngineImplTest.kt [hunk 3]
```diff
@@ -592,6 +595,87 @@ class FhirEngineImplTest {
...
+          id = "original-002"
+          meta =
+            Meta().apply {
+              versionId = "1"
+              lastUpdated = Date.from(Instant.parse("2022-12-02T10:15:30.00Z"))
+            }
+          addName(
+            HumanName().apply {
+              family = "Stark"
+              addGiven("Tony")
```
Comment (jingtang10) [inline engine/src/test/java/com/google/android/fhir/impl/FhirEngineImplTest.kt:612]: lol

File: engine/src/test/java/com/google/android/fhir/impl/FhirEngineImplTest.kt [hunk 4]
```diff
@@ -592,6 +595,87 @@ class FhirEngineImplTest {
...
+            },
+          )
+        }
+      // First sync
+      fhirEngine.syncDownload(AcceptLocalConflictResolver) { flowOf((listOf((originalPatient)))) }
+
+      val remoteChange =
```
Comment (jingtang10) [inline engine/src/test/java/com/google/android/fhir/impl/FhirEngineImplTest.kt:619]:
```suggestion
      val updatedPatient =
```

File: engine/src/test/java/com/google/android/fhir/impl/FhirEngineImplTest.kt [hunk 5]
```diff
@@ -592,6 +595,87 @@ class FhirEngineImplTest {
...
+          )
+        }
+      // First sync
+      fhirEngine.syncDownload(AcceptLocalConflictResolver) { flowOf((listOf((originalPatient)))) }
+
+      val localChange =
+        originalPatient.copy().apply { addAddress(Address().apply { city = "Malibu" }) }
+      fhirEngine.update(localChange)
+
+      val remoteChange =
```
Comment (jingtang10) [inline engine/src/test/java/com/google/android/fhir/impl/FhirEngineImplTest.kt:662]:
```suggestion
      val updatedPatient =
```

File: engine/src/main/java/com/google/android/fhir/db/impl/dao/ResourceDao.kt [hunk 6]
```diff
@@ -58,42 +58,57 @@ internal abstract class ResourceDao {
...
+          lastUpdatedRemote = lastUpdatedRemote?.toInstant(),
+          versionId = resource.versionId,
+        )
+      updateChanges(entity, resource)
     }
       ?: throw ResourceNotFoundException(resource.resourceType.name, resource.id)
   }
 
+  // WARNING : This is a private function and shouldn't be called directly.
+  open suspend fun updateChanges(entity: ResourceEntity, resource: Resource) {
```
Comment (jingtang10) [inline engine/src/main/java/com/google/android/fhir/db/impl/dao/ResourceDao.kt:88]: do these have to be open?

File: engine/src/main/java/com/google/android/fhir/db/impl/dao/ResourceDao.kt [hunk 7]
```diff
@@ -58,42 +58,57 @@ internal abstract class ResourceDao {
   lateinit var iParser: IParser
   lateinit var resourceIndexer: ResourceIndexer
 
-  open suspend fun update(resource: Resource, timeOfLocalChange: Instant?) {
+  open suspend fun updateLocalChange(resource: Resource, timeOfLocalChange: Instant?) {
```
[the remark this answers is not in the document]
Comment (aditya-07) [reply engine/src/main/java/com/google/android/fhir/db/impl/dao/ResourceDao.kt:61]: It is the same as the last downloaded remote version.

File: engine/src/main/java/com/google/android/fhir/db/impl/dao/ResourceDao.kt [hunk 8]
```diff
@@ -58,42 +58,57 @@ internal abstract class ResourceDao {
...
-            }
-          }
-          .build()
-      updateIndicesForResource(index, resource.resourceType, it.resourceUuid)
+      updateChanges(entity, resource)
+    }
+      ?: throw ResourceNotFoundException(resource.resourceType.name, resource.id)
+  }
+
+  open suspend fun updateRemoteChange(resource: Resource) {
```
[the remark this answers is not in the document]
Comment (aditya-07) [reply engine/src/main/java/com/google/android/fhir/db/impl/dao/ResourceDao.kt]: Done

File: engine/src/main/java/com/google/android/fhir/db/impl/dao/ResourceDao.kt [hunk 9]
```diff
@@ -58,42 +58,57 @@ internal abstract class ResourceDao {
...
+    getResourceEntity(resource.logicalId, resource.resourceType)?.let {
+      val lastUpdatedRemote: Date? = resource.meta.lastUpdated
```
[the remark this answers is not in the document]
Comment (aditya-07) [reply engine/src/main/java/com/google/android/fhir/db/impl/dao/ResourceDao.kt]: done

File: engine/src/main/java/com/google/android/fhir/db/impl/dao/ResourceDao.kt [hunk 10]
```diff
@@ -58,42 +58,67 @@ internal abstract class ResourceDao {
...
+            addDateTimeIndex(
+              createLastUpdatedIndex(resource.resourceType, InstantType(Date.from(instant))),
+            )
+          }
+        }
+        .build()
+    updateIndicesForResource(index, resource.resourceType, entity.resourceUuid)
+  }
+
   open suspend fun insertAllRemote(resources: List<Resource>): List<UUID> {
```
Comment (jingtang10) [inline engine/src/main/java/com/google/android/fhir/db/impl/dao/ResourceDao.kt:122]:
```suggestion
 suspend fun insertAllRemote(resources: List<Resource>): List<UUID> {
```

File: engine/src/main/java/com/google/android/fhir/db/impl/dao/ResourceDao.kt [hunk 11]
```diff
@@ -58,42 +58,67 @@ internal abstract class ResourceDao {
   lateinit var iParser: IParser
   lateinit var resourceIndexer: ResourceIndexer
 
-  open suspend fun update(resource: Resource, timeOfLocalChange: Instant?) {
+  /**
+   * Updates the resource in the [ResourceEntity] and adds indexes.
+   *
+   * @param [resource] The resource with local (on device) updates.
```
Comment (jingtang10) [inline engine/src/main/java/com/google/android/fhir/db/impl/dao/ResourceDao.kt:64]:
```suggestion
   * @param [resource] the resource with local (on device) updates
```

File: engine/src/main/java/com/google/android/fhir/db/impl/dao/ResourceDao.kt [hunk 12]
```diff
@@ -58,42 +58,67 @@ internal abstract class ResourceDao {
...
+   * @param [timeOfLocalChange] Time when the local change was made.
```
Comment (jingtang10) [inline engine/src/main/java/com/google/android/fhir/db/impl/dao/ResourceDao.kt:65]:
```suggestion
   * @param [timeOfLocalChange] time when the local change was made
```

File: engine/src/main/java/com/google/android/fhir/db/impl/dao/ResourceDao.kt [hunk 13]
```diff
@@ -58,42 +58,67 @@ internal abstract class ResourceDao {
...
-      updateIndicesForResource(index, resource.resourceType, it.resourceUuid)
+      updateChanges(entity, resource)
     }
       ?: throw ResourceNotFoundException(resource.resourceType.name, resource.id)
   }
 
+  /**
+   * Updates the resource in the [ResourceEntity] and adds indexes.
+   *
+   * @param [resource] The resource with the remote(server) updates.
```
Comment (jingtang10) [inline engine/src/main/java/com/google/android/fhir/db/impl/dao/ResourceDao.kt:82]:
```suggestion
   * @param [resource] the resource with the remote(server) updates
```

File: engine/src/main/java/com/google/android/fhir/db/impl/dao/ResourceDao.kt [hunk 14]
```diff
@@ -58,42 +58,67 @@ internal abstract class ResourceDao {
...
-          }
-          .build()
-      updateIndicesForResource(index, resource.resourceType, it.resourceUuid)
+      updateChanges(entity, resource)
     }
       ?: throw ResourceNotFoundException(resource.resourceType.name, resource.id)
   }
 
+  /**
+   * Updates the resource in the [ResourceEntity] and adds indexes.
```
Comment (jingtang10) [inline engine/src/main/java/com/google/android/fhir/db/impl/dao/ResourceDao.kt:80]:
```suggestion
   * Updates the resource in the [ResourceEntity] and adds indexes as a result of downloading the resource from server.
```

File: engine/src/main/java/com/google/android/fhir/db/impl/dao/ResourceDao.kt [hunk 15]
```diff
@@ -58,42 +58,67 @@ internal abstract class ResourceDao {
   lateinit var iParser: IParser
   lateinit var resourceIndexer: ResourceIndexer
 
-  open suspend fun update(resource: Resource, timeOfLocalChange: Instant?) {
+  /**
+   * Updates the resource in the [ResourceEntity] and adds indexes.
```
Comment (jingtang10) [inline engine/src/main/java/com/google/android/fhir/db/impl/dao/ResourceDao.kt:62]:
```suggestion
   * Updates the resource in the [ResourceEntity] and adds indexes as a result of changes made on device.
```

File: engine/src/main/java/com/google/android/fhir/db/impl/dao/ResourceDao.kt [hunk 16]
```diff
@@ -58,42 +58,67 @@ internal abstract class ResourceDao {
...
+          serializedResource = iParser.encodeResourceToString(resource),
+          lastUpdatedRemote = resource.meta.lastUpdated?.toInstant(),
+          versionId = resource.versionId,
+        )
+      updateChanges(entity, resource)
+    }
+      ?: throw ResourceNotFoundException(resource.resourceType.name, resource.id)
+  }
+
+  // WARNING : This is a private function and shouldn't be called directly.
```
Comment (jingtang10) [inline engine/src/main/java/com/google/android/fhir/db/impl/dao/ResourceDao.kt:97]:
```suggestion
```

**Review by jingtang10 — APPROVED** [review]

File: engine/src/main/java/com/google/android/fhir/db/impl/dao/ResourceDao.kt (hunk 6 above)
Comment (aditya-07) [reply engine/src/main/java/com/google/android/fhir/db/impl/dao/ResourceDao.kt]: Removed open and made it private.

**Pull request merged**